tests/test_asyncio.py

```diff
@@ -30,34 +31,125 @@ async def test_asyncio_bad_put_raises(dummy_server_async):
+        await asyncio.sleep(0.2)
+        flush_event.set()
+        return await data_generator.__anext__()
+
+    async with AsyncioClient("localhost") as client:
+        data_generator = client.data(
+            frame_timeout=5,
+            flush_event=flush_event,
+            flush_mode=FlushMode.MANUAL,
+        )
+
+        assert isinstance(await data_generator.__anext__(), ReadyData)
+        assert isinstance(await data_generator.__anext__(), StartData)
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert isinstance(await data_generator.__anext__(), EndData)
+        await data_generator.aclose()
+
+
+async def test_asyncio_data_flush_every_frame(
+    dummy_server_async, fast_dump, fast_dump_expected
+):
     dummy_server_async.data = fast_dump
+
     events = []
     async with AsyncioClient("localhost") as client:
-        async for data in client.data(frame_timeout=1, flush_period=flush_period):
+        async for data in client.data(
+            frame_timeout=5,
+            flush_mode=FlushMode.IMMEDIATE,
+        ):
             events.append(data)
+
             if len(events) == len(fast_dump_expected):
                 break
+
     assert fast_dump_expected == events
 
 
+async def test_asyncio_data_timeout_on_no_manual_press(dummy_server_async, fast_dump):
+    flush_event = asyncio.Event()
+    async with AsyncioClient("localhost") as client:
+        with pytest.raises(asyncio.TimeoutError, match="No data received for 3s"):
+            async for data in client.data(
+                frame_timeout=3, flush_mode=FlushMode.MANUAL, flush_event=flush_event
+            ):
+                pass
+
+
 async def test_asyncio_data_timeout(dummy_server_async, fast_dump):
     dummy_server_async.data = fast_dump
     async with AsyncioClient("localhost") as client:
         with pytest.raises(asyncio.TimeoutError, match="No data received for 0.1s"):
-            async for data in client.data(frame_timeout=0.1):
+            async for data in client.data(
+                frame_timeout=0.1, flush_mode=FlushMode.IMMEDIATE
+            ):
                 "This goes forever, when it runs out of data we will get our timeout"
 
 
+async def test_asyncio_data_nonexistent_flushmode(dummy_server_async, fast_dump):
+    dummy_server_async.data = fast_dump
+    async with AsyncioClient("localhost") as client:
+        with pytest.raises(ValueError, match="flush_mode must be one of"):
+            async for data in client.data(frame_timeout=0.1, flush_mode=None):
+                pass
+
+
+@pytest.mark.parametrize(
```

Comment:
Worth adding a test for passing something that isn't a valid `FlushMode` enum value?

Reply:
Like this test but using something other than None?

https://github.com/PandABlocks/PandABlocks-client/pull/64/commits/f8b57de721d528abfb9f9a10682bf92662518545#diff-55234053c6f9538eec8b76e6fce1715585c5359a3a8362042d1041381c6bd751R129

Reply:
Apologies, I missed that one. That one is fine.